endpoints/s3/src/putobject.cpp

```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+				if (chunked_flag || special_chunked_header) {
+					// chunked - get part size from x-amz-decoded-content-length
+					// if this isn't provided then just continue without storing the part size
+					auto header = parser_message.find("x-amz-decoded-content-length");
+					if (header != parser_message.end()) {
+						part_size = boost::lexical_cast<uint64_t>(parser_message["x-amz-decoded-content-length"]);
+
+						// Make sure someone hasn't previously uploaded this part with a different part size.
+						// If so we can't handle that and have to reject the call.
+						if (part_shmem::part_size_map[upload_id].find(part_number_int) !=
+						    part_shmem::part_size_map[upload_id].end()) {
+							auto old_part_size = part_shmem::part_size_map[upload_id][part_number_int];
+							if (old_part_size != part_size) {
+								// reject this
+								logging::error(
+									"{}: Upload ID {} - part_number {} was uploaded a second time with a different "
+									"part size. Old part size = {}. New part size = {}. "
+									"Rejecting this request.",
+									__func__,
+									upload_id,
+									part_number_int,
+									old_part_size,
+									part_size);
+								response.result(beast::http::status::bad_request);
+								logging::debug("{}: returned {}", __func__, response.reason());
+								session_ptr->send(std::move(response));
+								return;
+							}
+						}
+
+						part_shmem::part_size_map[upload_id][part_number_int] = part_size;
+					}
+				}
+				else {
+					// not chunked, get part size from content_length
+					part_size = boost::lexical_cast<uint64_t>(parser_message[beast::http::field::content_length]);
+
+					// Make sure someone hasn't previously uploaded this part with a different part size.
+					// If so we can't handle that and have to reject the call.
+					if (part_shmem::part_size_map[upload_id].find(part_number_int) !=
+					    part_shmem::part_size_map[upload_id].end()) {
+						auto old_part_size = part_shmem::part_size_map[upload_id][part_number_int];
+						if (old_part_size != part_size) {
+							// reject this
+							logging::error(
+								"{}: Upload ID {} - part_number {} was uploaded a second time with a different part "
+								"size. "
+								"Rejecting this request",
+								__func__,
+								upload_id,
+								part_number_int);
+							response.result(beast::http::status::bad_request);
+							logging::debug("{}: returned {}", __func__, response.reason());
+							session_ptr->send(std::move(response));
+							return;
+						}
+					}
+
+					part_shmem::part_size_map[upload_id][part_number_int] = part_size;
+				}
+
+				// see if we know all of the previous part_sizes
+				know_part_offset = true;
+				for (unsigned int i = 1; i <= part_number_int - 1; ++i) {
+					if (part_shmem::part_size_map[upload_id].find(i) != part_shmem::part_size_map[upload_id].end()) {
+						part_offset += part_shmem::part_size_map[upload_id][i];
+					}
+					else {
+						know_part_offset = false;
```

Comment:
If we get into this case, can we break out of the loop? Will we ever need the `part_offset` to be accurate if any of the previous `part_size`s are not known? Or am I misunderstanding?

Reply:
We can break.  I made the change.